swarm/swap/swap.go

```diff
@@ -0,0 +1,347 @@
+//It then handles the event received, filtering for messages and evaluating
+//if it needs accounting
+func (s *Swap) registerForEvents(srv *p2p.Server) {
+	go func() {
+		events := make(chan *p2p.PeerEvent)
+		sub := srv.SubscribeEvents(events)
+		defer sub.Unsubscribe()
+
+		for {
+			select {
+			case event := <-events:
+				go s.handleMsgEvent(event)
+			case err := <-sub.Err():
+				log.Error(err.Error())
+				return
+			}
+		}
+	}()
+}
+
+//Handle the message.
+//Determine if it needs accounting, and if yes, account for it
+func (s *Swap) handleMsgEvent(event *p2p.PeerEvent) {
+	if !s.priceOracle.IsAccountedMsg(event) {
+		return
+	}
+	s.AccountMsgForPeer(event)
+}
+
+//Do the accounting
+//Depending on the charging type of the message (set in the PriceTag),
+//it will charge the sender or receiver
+func (s *Swap) AccountMsgForPeer(event *p2p.PeerEvent) {
+	price, direction := s.priceOracle.GetPriceForMsg(event)
+	if price == nil {
+		//TODO what to do in this case? Should not happen
+		log.Crit("Price is nil; this should have been accounted for but somehow it failed")
+	}
+
+	if direction == ChargeSender {
+		//we are sending a ChargeSender message, thus debit us and credit remote
+		if event.Type == p2p.PeerEventTypeMsgSend {
+			s.chargeLocal(event, price)
+			//we are receiving a ChargeSender message, so credit us and debit remote
+		} else if event.Type == p2p.PeerEventTypeMsgRecv {
+			s.chargeRemote(event, price)
+		}
+	} else {
+		//we are receiving a ChargeReceiver message, thus debit us and credit remote
+		if event.Type == p2p.PeerEventTypeMsgRecv {
+			s.chargeLocal(event, price)
+			//we are sending a ChargeReceiver message, thus credit us and debit remote
+		} else if event.Type == p2p.PeerEventTypeMsgSend {
+			s.chargeRemote(event, price)
+		}
+	}
+}
+
+//Debit us and credit remote
+func (s *Swap) chargeLocal(event *p2p.PeerEvent, amount *big.Int) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	//local node is being credited (in its favor), so its balance increases
+	if s.peers[event.Peer] == nil {
+		s.peers[event.Peer] = &big.Int{}
+		s.stateStore.Get(event.Peer.String(), s.peers[event.Peer])
+	}
+	peerBalance := s.peers[event.Peer]
+	peerBalance.Sub(peerBalance, amount)
+	//local node is being debited (in favor of remote peer), so its balance decreases
+	//TODO: save to store here? init store?
+	s.stateStore.Put(event.Peer.String(), peerBalance)
+	//(balance *Int) Cmp(payAt)
+	// -1 if balance <  payAt
+	//  0 if balance == payAt
+	// +1 if balance >  payAt
+	if peerBalance.Cmp(payAt) == -1 {
+		ctx := context.TODO()
+		err := s.issueCheque(ctx, event.Peer)
+		if err != nil {
+			//TODO: special error handling, as at this point the accounting has been done
+			//but the cheque could not be sent?
+			log.Warn("Payment threshold exceeded, but error sending cheque!", "err", err)
+		}
+	}
+	if peerBalance.Cmp(dropAt) == -1 {
+		peer := s.protocol.getPeer(event.Peer)
+		if peer != nil {
+			peer.Drop(ErrInsufficientFunds)
```

Comment:
`Disconnect` can just be called on `event.Peer` i think

Reply:
`event.Peer` is just an ID, so we can not call `Disconnect` on it

Reply:
ok